core/src/main/java/io/grpc/internal/DnsNameResolver.java

```diff
@@ -501,86 +502,110 @@ ResolutionResults resolve(String host) throws NamingException {
     }
   }
 
-  @VisibleForTesting
-  static JsonArray parseTxtResults(List<String> txtRecords) {
-    JsonArray serviceConfigs = new JsonArray();
-    Gson gson = new Gson();
-
-    for (String txtRecord : txtRecords) {
-      if (txtRecord.startsWith(SERVICE_CONFIG_PREFIX)) {
-        JsonArray choices;
-        try {
-          choices =
-              gson.fromJson(txtRecord.substring(SERVICE_CONFIG_PREFIX.length()), JsonArray.class);
-        } catch (JsonSyntaxException e) {
-          logger.log(Level.WARNING, "Bad service config: " + txtRecord, e);
-          continue;
-        }
-        serviceConfigs.addAll(choices);
-      } else {
-        logger.log(Level.FINE, "Ignoring non service config {0}", new Object[]{txtRecord});
-      }
-    }
-
-    return serviceConfigs;
-  }
-
   /**
    * Determines if a given Service Config choice applies, and if so, returns it.
    *
    * @see <a href="https://github.com/grpc/proposal/blob/master/A2-service-configs-in-dns.md">
-   *   Service Config in DNS</a>
-   * @param choice The service config choice.
+   *     Service Config in DNS</a>
    * @return The service config object or {@code null} if this choice does not apply.
    */
   @VisibleForTesting
   @Nullable
-  @SuppressWarnings("BetaApi") // Verify isn't all that beta
-  static JsonObject maybeChooseServiceConfig(JsonObject choice, Random random, String hostname) {
-    for (Entry<String, ?> entry : choice.entrySet()) {
-      Verify.verify(SERVICE_CONFIG_CHOICE_KEYS.contains(entry.getKey()), "Bad key: %s", entry);
-    }
-    if (choice.has("clientLanguage")) {
-      JsonArray clientLanguages = choice.get("clientLanguage").getAsJsonArray();
-      if (clientLanguages.size() != 0) {
-        boolean javaPresent = false;
-        for (JsonElement clientLanguage : clientLanguages) {
-          String lang = clientLanguage.getAsString().toLowerCase(Locale.ROOT);
-          if ("java".equals(lang)) {
-            javaPresent = true;
-            break;
+  @SuppressWarnings("unchecked")
+  static Map<String, Object> maybeChooseServiceConfig(
+      JsonReader reader, Random random, String hostname) throws ClassCastException, IOException {
+    Map<String, Object> serviceConfig = null;
+    reader.beginObject();
```

Comment:
I'd really rather we didn't mix parsing with business logic. Can we just `readJsonElement()` and the process the resulting object? Then we can use things like `get("clientLanguage")`.